[GStreamer] Buffer end equals video duration for progressive video

eocanha opened the pull request:
#### 3e214abe945921fb2ebbf234c7bfd1dab35cea9e
<pre>
[GStreamer] Buffer end equals video duration for progressive video
<a href="https://bugs.webkit.org/show_bug.cgi?id=269203">https://bugs.webkit.org/show_bug.cgi?id=269203</a>

Reviewed by Philippe Normand.

On some long regular videos, the current buffer management mechanism
reports 100% of the video as loaded when that&apos;s not actually happening.
This causes document.getElementsByTagName(&apos;video&apos;)[0].buffered.end(0) to
be equal to document.getElementsByTagName(&apos;video&apos;)[0].duration when that
should clearly not the case.

The current default way of computing m_maxTimeLoaded isn&apos;t working
properly because oftern, when the buffering messages are received by
MediaPlayerPrivateGStreamer, the duration is still unknown and the code
in charge of computing the fill status bails out.

The WebKitWebSrc network downloading statistics provide a much more
accurate source of information for the buffered data, and should be used
in more cases when available. The download statistics tend to arrive
later than the buffer fill messages and at that moment duration is
already available, leading to a proper computation of m_maxTimeLoaded.

See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1238">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1238</a>

This patch calls updateMaxTimeLoaded() when the network statistics are
received. That update must be done here because the current system in
place to perform such update is based on m_fillTimer, and when the
network stats are received the timer is usually already stopped. This
method works fine both with on-disk cache and without it (the typical
use case on set-top-box devices).

Original patch authored by: suresh-khurdiya-epam &lt;skhurdiya.contractor@libertyglobal.com&gt;

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleMessage): When a webkit-network-statistics message is received, compute the fill status and update m_maxTimeLoaded.

Canonical link: <a href="https://commits.webkit.org/274643@main">https://commits.webkit.org/274643@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ffe6ebbba7865fc819ea6c20cdf91f67806a09aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39640 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18619 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41996 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42175 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35540 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41946 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21518 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15948 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33094 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40214 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15709 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34305 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13621 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13607 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35259 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43452 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35999 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35592 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39382 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14452 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11906 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37649 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16058 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8885 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16107 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15715 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8a1661241d96539da722a27248b429d0bae821aa. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24257

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2006,8 +2006,18 @@ void MediaPlayerPrivateGStreamer::handleMessage(GstMessage* message)
+                if (mediaDuration && m_httpResponseTotalSize) {
+                    const double fillStatus = 100.0 * (static_cast<double>(m_networkReadPosition) / static_cast<double>(m_httpResponseTotalSize));
+                    updateMaxTimeLoaded(fillStatus);
+                    GST_DEBUG("Updated maxTimeLoaded base on network read position: %s", toString(m_maxTimeLoaded).utf8().data());
```
That conversion doesn't seem needed.

eocanha replied:
I've already clarified to you in private that the conversion to `String` is needed, because `m_maxTimeLoaded` is a `MediaTime`, not a `double` or anything else trivially convertible.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ffe6ebbba7865fc819ea6c20cdf91f67806a09aa. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24257

eocanha committed:
[GStreamer] Buffer end equals video duration for progressive video
https://bugs.webkit.org/show_bug.cgi?id=269203

Reviewed by Philippe Normand.

On some long regular videos, the current buffer management mechanism
reports 100% of the video as loaded when that's not actually happening.
This causes document.getElementsByTagName('video')[0].buffered.end(0) to
be equal to document.getElementsByTagName('video')[0].duration when that
should clearly not the case.

The current default way of computing m_maxTimeLoaded isn't working
properly because oftern, when the buffering messages are received by
MediaPlayerPrivateGStreamer, the duration is still unknown and the code
in charge of computing the fill status bails out.

The WebKitWebSrc network downloading statistics provide a much more
accurate source of information for the buffered data, and should be used
in more cases when available. The download statistics tend to arrive
later than the buffer fill messages and at that moment duration is
already available, leading to a proper computation of m_maxTimeLoaded.

See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1238

This patch calls updateMaxTimeLoaded() when the network statistics are
received. That update must be done here because the current system in
place to perform such update is based on m_fillTimer, and when the
network stats are received the timer is usually already stopped. This
method works fine both with on-disk cache and without it (the typical
use case on set-top-box devices).

Original patch authored by: suresh-khurdiya-epam <skhurdiya.contractor@libertyglobal.com>

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::handleMessage): When a webkit-network-statistics message is received, compute the fill status and update m_maxTimeLoaded.

Canonical link: https://commits.webkit.org/274643@main

webkit-commit-queue commented:
Committed 274643@main (3e214abe9459): <https://commits.webkit.org/274643@main>

Reviewed commits have been landed. Closing PR #24257 and removing active labels.
